sfge/src/main/java/com/salesforce/rules/ops/boundary/SharingPolicyBoundary.java

```diff
@@ -0,0 +1,130 @@
+package com.salesforce.rules.ops.boundary;
+
+import com.salesforce.graph.Schema;
+import com.salesforce.graph.vertex.SFVertex;
+import com.salesforce.graph.vertex.SFVertexFactory;
+import com.salesforce.graph.vertex.SharingPolicy;
+import com.salesforce.graph.vertex.UserClassVertex;
+import java.util.List;
+import java.util.Optional;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__;
+
+/**
+ * This class stores a {@link UserClassVertex} boundary item and can try to find the nearest
+ * ancestor with an explicitly defined sharing policy (the "applicable ancestor"). See {@link
+ * #getApplicableAncestorPolicyVertex()}.
+ */
+public class SharingPolicyBoundary implements Boundary<SFVertex> {
+    private final GraphTraversalSource g;
+    private final UserClassVertex boundaryItem;
+    /**
+     * stores the applicable ancestor, containing a sharing policy (or omitted sharing policy) that
+     * applies to the boundary item. See {@link #getApplicableAncestorPolicyVertex()}.
+     */
+    private UserClassVertex applicableAncestorVertex;
+
+    /**
+     * store the {@link UserClassVertex} associated with this boundary.
+     *
+     * @param boundaryItem the {@link UserClassVertex} that the boundary represents
+     * @param g a {@link GraphTraversalSource} used to find the applicable ancestor. See {@link
+     *     #getApplicableAncestorPolicyVertex()}.
+     */
+    public SharingPolicyBoundary(UserClassVertex boundaryItem, GraphTraversalSource g) {
+        this.boundaryItem = boundaryItem;
+        this.g = g;
+        // save the work of recursively querying the graph until it's required
+        this.applicableAncestorVertex = null;
+    }
+
+    /**
+     * @return the {@link UserClassVertex} boundary item
+     */
+    @Override
+    public UserClassVertex getBoundaryItem() {
+        return this.boundaryItem;
+    }
+
+    /**
+     * Try to find the applicable ancestor to the boundary item. The applicable ancestor is defined
```

Comment:
"Ancestor" feels ambiguous. Is there anything less ambiguous you could use instead? I'm thinking "superclass".